### PR TITLE
[performance](memtable) do not copy input_block if column matches

### PR DESCRIPTION
**Pull request by kaijchen**
## Proposed changes

In most common cases, columns in `slot_descs` and `tuple_desc->slots()` will match and we don't need to convert `input_block` in `MemTable::insert()`.

For materialized views, `input_block` will be converted by `_column_offset` as intended, see #9923.

To optimize performance in cases with many (e.g. 12k+) tablets:

Before:

```
-  DeltaWriterWriteTime:  1m58s
   -  MemTableWriterFlushTime:  0ns
   -  MemTableWriterLockTime:  1s41ms
   -  MemTableWriterShrinkTime:  0ns
   -  MemTableWriterWriteTime:  1m53s
      -  MemTableCopyCount:  12.044003M  (12044003)
      -  MemTableCopyTime:  42s857ms
      -  MemTableCopyTime0:  11s866ms
      -  MemTableInitTime:  13.384ms
      -  MemTableInsertTime:  51s543ms
```

After:

```
-  DeltaWriterWriteTime:  1m
   -  MemTableWriterFlushTime:  0ns
   -  MemTableWriterLockTime:  971.481ms
   -  MemTableWriterShrinkTime:  0ns
   -  MemTableWriterWriteTime:  55s274ms
      -  MemTableCopyCount:  0
      -  MemTableCopyTime:  0ns
      -  MemTableCopyTime0:  0ns
      -  MemTableInitTime:  18.746ms
      -  MemTableInsertTime:  53s772ms
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kaijchen** [comment]
run buildall

File: be/src/olap/memtable.h [hunk 1]
```diff
@@ -231,6 +231,7 @@ class MemTable {
     void _init_columns_offset_by_slot_descs(const std::vector<SlotDescriptor*>* slot_descs,
                                             const TupleDescriptor* tuple_desc);
     std::vector<int> _column_offset;
+    bool _column_match = true;
```
Comment (eldenmoon) [inline be/src/olap/memtable.h:234]: add comment to explain this variable

Comment (kaijchen) [inline be/src/olap/memtable.h:234]:
```suggestion
    // Wheather columns in input block matches with columns in tablet.
    // Usually `true` for base tables, `false` for materialized views.
    // This field will be set during `_init_columns_offset_by_slot_descs`.
    bool _column_match = true;
```

File: be/src/olap/memtable.h [hunk 2]
```diff
@@ -231,6 +231,10 @@ class MemTable {
     void _init_columns_offset_by_slot_descs(const std::vector<SlotDescriptor*>* slot_descs,
                                             const TupleDescriptor* tuple_desc);
     std::vector<int> _column_offset;
+    // Wheather columns in input block matches with columns in tablet.
```
Comment (kaijchen) [inline be/src/olap/memtable.h:234]:
```suggestion
    // Whether columns in input block matches with columns in tablet.
```

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Review by liaoxin01 — DISMISSED** [review]
LGTM

**Comment by liaoxin01** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
Superseded by #36939

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
Superseded by #36939

**Pull request closed without merging**